Mage.Sets/src/mage/cards/a/AminatousAugury.java

```diff
@@ -134,54 +136,60 @@ public AminatousAuguryCastFromExileEffect copy() {
     }
 
     @Override
-    public boolean applies(UUID sourceId, Ability source, UUID affectedControllerId, Game game) {
-        Player player = game.getPlayer(affectedControllerId);
-        EnumSet<CardType> usedCardTypes = EnumSet.noneOf(CardType.class);
+    public boolean applies(UUID objectId, Ability source, UUID affectedControllerId, Game game) {
+        return applies(objectId, null, source, game, affectedControllerId);
+    }
+
+
+    @Override
+    public boolean applies(UUID objectId, Ability affectedAbility, Ability source, Game game, UUID playerId) {
+
+        Card cardToCheck = game.getCard(objectId);
+        objectId = CardUtil.getMainCardId(game, objectId); // for split cards
 
-        if (game.getState().getValue(source.getSourceId().toString() + "cardTypes") != null) {
-            usedCardTypes = (EnumSet<CardType>) game.getState().getValue(source.getSourceId().toString() + "cardTypes");
+        if (!isAbilityAppliedForAlternateCast(cardToCheck, affectedAbility, playerId, source)) {
+            return false;
         }
-        //TODO add code for adding additional costs to the card
-        if (player != null
-                && sourceId != null
-                && sourceId.equals(getTargetPointer().getFirst(game, source))
-                && affectedControllerId.equals(source.getControllerId())) {
-            Card card = game.getCard(sourceId);
-            if (card != null
-                    && game.getState().getZone(sourceId) == Zone.EXILED) {
-                EnumSet<CardType> unusedCardTypes = EnumSet.noneOf(CardType.class);
-                for (CardType cardT : card.getCardType()) {
-                    if (!usedCardTypes.contains(cardT)) {
-                        unusedCardTypes.add(cardT);
-                    }
+
+        //cards put into exile using Aminatou -- created in AminatouAuguryEffect - Apply
+        Set<Card> aminatouCards = (Set<Card>) game.getState().getValue(source.getSourceId().toString() + "Cards");
+
+        //Set to track which card types from the Aminatou cards have already been cast
```

Comment:
Applies method calls multiple times to check playable cards or allows actions. You can't setup used mark here (as example: it will be change on one check but other checks will be skips). Bugs example: if you can multiple cards from exile (with same or different types). Also you can't use any choose dialogs here.

I recommends to more research. Hmm.